runtime/lua/vim/lsp/semantic_tokens.lua

```diff
@@ -0,0 +1,164 @@
+local M = {}
+
+local semantic_tokens = setmetatable({}, {
+  __index = function(table, key)
+    local rtn = {}
+    rawset(table, key, rtn)
+    return rtn
+  end,
+})
+local last_tick = {}
+local active_requests = {}
+
+---@private
+local function get_bit(n, k)
+  --todo(theHamsta): remove once `bit` module is available for non-LuaJIT
+  if _G.bit then
+    return _G.bit.band(_G.bit.rshift(n, k), 1)
+  else
+    return (n / math.pow(2, k)) % 2
+  end
+end
+
+---@private
+local function modifiers_from_number(x, modifiers_table)
+  local modifiers = {}
+  for i = 0, #modifiers_table - 1 do
+    local bit = get_bit(x, i)
+    if bit == 1 then
+      table.insert(modifiers, 1, modifiers_table[i + 1])
+    end
+  end
+
+  return modifiers
+end
+
+--- |lsp-handler| for the method `textDocument/semanticTokens/full`
+---
+--- This function can be configured with |vim.lsp.with()| with the following options for `config`
+---
+--- `on_token`: A function with signature `function(ctx, token)` that is called
+---             whenever a semantic token is received from the server from context `ctx`
+---             (see |lsp-handler| for the definition of `ctx`). This can be used for highlighting the tokens.
+---             `token` is a table:
+---
+--- <pre>
+---   {
+---         line             -- line number 0-based
+---         start_char       -- start character 0-based
+---         length           -- length in characters of this token
+---         type             -- token type as string (see https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide#semantic-token-classification)
+---         modifiers        -- token modifier as string (see https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide#semantic-token-classification)
+---         offset_encoding  -- offset encoding used by the language server (see |lsp-sync|)
+---   }
+--- </pre>
+---
+--- `on_invalidate_range`: A function with signature `function(ctx, line_start, line_end)` called whenever tokens
+---                        in a specific line range (`line_start`, `line_end`) should be considered invalidated
+---                        (see |lsp-handler| for the definition of `ctx`). `line_end` can be -1 to
+---                        indicate invalidation until the end of the buffer.
+function M.on_full(err, response, ctx, config)
+  if not ctx.bufnr or not ctx.client_id then
+    return
+  end
+  active_requests[ctx.bufnr] = false
+  if config and config.on_invalidate_range then
+    config.on_invalidate_range(ctx, 0, -1)
```

Comment:
One issue with this is that ctx holds the client id, which hypothetically could refer to a nil'd client if the client shuts down after receiving, since the client check is done later on line 73-76

Reply:
It's more a question of who we want to be responsible. This shfits the burden onto the users when they pass an on_invalidate_range callback to first check the client_id is not null. If we check for valid clients beforehand then it doesn't allow users to customize the behavior since the callback will never be called.

Your call, just wanted to raise the point.

Reply:
The client id will still be valid, the client object will not. This was one of my concerns (cleanup after client shutdown) which I mentioned in the other comment.

Reply:
Still need to check the client pointed by client_id is not nil

Reply:
But what would you do in that case? Not call the callback? Nil-fy `ctx.client_id`? The interface for the callbacks is more when do tokens appear and when they should be considered invalidated. It would be possible to do the check for client being nil but then one would probably want to register this function on client shutdown

Reply:
A now I get it you mean to cancel the whole function, right? Will update that

Reply:
@mjlbach I interpreted your comment that invalid client id should not trigger invalidation. There was already a check for the client and it issued a last invalidate. I moved it now further up and invalid client will now result in no-op-

Reply:
Cool, thanks!